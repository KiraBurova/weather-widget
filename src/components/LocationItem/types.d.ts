import { DraggableProvidedDragHandleProps } from 'react-beautiful-dnd';
import { TLocation } from '../../types';

export interface LocationItemProps {
  location: TLocation;
  dragHandleProps?: DraggableProvidedDragHandleProps;
}
