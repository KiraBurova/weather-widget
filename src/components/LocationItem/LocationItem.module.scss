@import '../../styles/variables';

.item {
  display: flex;
  align-items: center;
  background-color: $main-dark;
  border-radius: $border-radius;
  font-size: 1rem;
  padding: 0.5em;
  text-transform: capitalize;
}

.dragIcon {
  font-size: 24px;
  cursor: pointer;
  margin-right: 10px;
}
