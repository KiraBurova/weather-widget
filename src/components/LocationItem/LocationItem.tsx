import React from 'react';

import { Icon } from '@iconify/react';
import DragIcon from '@iconify/icons-ic/outline-drag-handle';

import styles from './LocationItem.module.scss';
import { LocationItemProps } from './types';

const LocationItem = ({ location }: LocationItemProps) => {
  return (
    <div className={styles.item}>
      <Icon icon={DragIcon} className={styles.dragIcon} />
      {location.name}
    </div>
  );
};

export default LocationItem;
