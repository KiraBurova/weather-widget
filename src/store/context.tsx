import { useLocalObservable } from 'mobx-react-lite';
import React from 'react';

import { createStore, TStore } from './store';

const StoreContext = React.createContext<TStore | null>(null);

type StoreProviderProps = {
  children: React.ReactNode;
};

export const StoreProvider = ({ children }: StoreProviderProps) => {
  const store = useLocalObservable(createStore);
  return <StoreContext.Provider value={store}>{children}</StoreContext.Provider>;
};

export const useStore = () => {
  const store = React.useContext(StoreContext);
  if (!store) {
    throw new Error('useStore must be used within a StoreProvider.');
  }
  return store;
};
