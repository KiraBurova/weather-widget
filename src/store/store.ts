import { getWeatherReportByCityName, getWeatherReportByLocation, getWeatherIconByIconName } from '../api';

import { TLocation, WeatherReportType } from '../types';

export function createStore() {
  return {
    location: '',
    locations: [] as TLocation[],
    weatherReports: [] as WeatherReportType[],
    fetchWeatherReportByLocation(latitude: number, longitude: number) {
      getWeatherReportByLocation(latitude, longitude)
        .then((weatherData: any) => {
          console.log(weatherData);
          const w = this.transformLocationData(weatherData.data);
          this.setWeatherReports(w);
        })
        .catch((error: Error) => console.log(error));
    },
    fetchWeatherReportByCityName(location: TLocation) {
      getWeatherReportByCityName(location.name)
        .then((weatherData) => {
          this.setLocations(location);
          const w = this.transformLocationData(weatherData.data);
          this.setWeatherReports(w);
        })
        .catch((error: Error) => console.log(error));
    },
    setLocations(location: TLocation) {
      this.locations.push(location);
    },
    deleteLocation(id: number) {
      this.locations = this.locations.filter((location) => location.id !== id);
    },
    setWeatherReports(weatherReport: any) {
      this.weatherReports.push(weatherReport);
    },
    transformLocationData(weatherData: WeatherReportType) {
      return {
        cityName: weatherData.name,
        countryName: weatherData.sys.country,
        weatherDescription: weatherData.weather[0].description,
        feelsLike: weatherData.main.feels_like,
        weatherIcon: getWeatherIconByIconName(weatherData.weather[0].icon),
        temperature: weatherData.main.temp,
        wind: weatherData.wind,
        humidity: weatherData.main.humidity,
        visibility: weatherData.visibility,
        pressure: weatherData.main.pressure,
        clouds: weatherData.clouds.all,
      };
    },
  };
}
export type TStore = ReturnType<typeof createStore>;
