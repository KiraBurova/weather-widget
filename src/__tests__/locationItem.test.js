import React from 'react';

import { render } from '@testing-library/react';

import store from '../store/store';
import { StoreProvider } from '../store/context';

import LocationItem from '../components/LocationItem';

describe('<LocationItem />', () => {
  it('renders the <LocationItem /> width given data', () => {
    const location = {
      name: 'Oslo',
      id: 2,
    };
    const { container, getByRole } = render(
      <StoreProvider rootStore={store}>
        <LocationItem location={location}></LocationItem>
      </StoreProvider>,
    );
    expect(getByRole('button')).toBeTruthy();

    expect(container.firstChild).toMatchSnapshot();
  });
});
