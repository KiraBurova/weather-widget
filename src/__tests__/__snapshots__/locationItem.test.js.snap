// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LocationItem /> renders the <LocationItem /> width given data 1`] = `
<div
  class="item"
>
  <div>
    <svg
      class="dragIcon"
      focusable="false"
      height="1em"
      preserveAspectRatio="xMidYMid meet"
      style="transform: rotate(360deg);"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M20 9H4v2h16V9zM4 15h16v-2H4v2z"
        fill="currentColor"
      />
    </svg>
  </div>
  Oslo
  <span
    class="deleteIcon"
    role="button"
  >
    <svg
      focusable="false"
      height="1em"
      preserveAspectRatio="xMidYMid meet"
      style="transform: rotate(360deg);"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM8 9h8v10H8V9zm7.5-5l-1-1h-5l-1 1H5v2h14V4h-3.5z"
        fill="currentColor"
      />
    </svg>
  </span>
</div>
`;
